Added "val" attribute to widgets.RadioButtons

DanHickstein opened the pull request:
It would be nice to be able to query the current selection of a set of RadioButtons using "rButton.val", just like one can query the current value of a slider using "mySlider.val". I included two lines of code that allow this to be done.

tacaswell commented:
Thanks, looks good!  

Can you squash this down to one commit?

I don't think there are currently any tests of the this widget, but if you want to take a crack at adding them it would be awesome (check out the selector tests for a template).

DanHickstein commented:
Sounds good.

But I am a new to using github. How do I make it just one commit? Do I need
to fork and make the changes again?

On Mon, Jan 5, 2015 at 7:03 PM, Thomas A Caswell <notifications@github.com>
wrote:

> Thanks, looks good!
>
> Can you squash this down to one commit?
>
> I don't think there are currently any tests of the this widget, but if you
> want to take a crack at adding them it would be awesome (check out the
> selector tests for a template).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/3971#issuecomment-68815536>
> .
>



-- 
Daniel D. Hickstein
DanHickstein@gmail.com
Phone: 301 787 1853

tacaswell commented:
It is best-practice to make changes on a branch _other_ than master.  To squash you need to do an 'interactive rebase'.  How familiar with git are you and how patient are you feeling tonight? ;)

The very short version of how to do this is to change to the directory where have the repo checked out and do

```
git rebase -i upstream/master
```
where I am assuming that  the remote 'upstream' points towards the canonical mpl repo, and follow the on-screen directions.  If you use a GUI to interact with git any reasonable one should have the ability to do this hidden somewhere.

You then do a:

```
git push --force origin master:master
```

where I am assuming that the remote 'origin' is _your_ mpl repo on github (you should try this command _without_ the force to see what it does, in general `--force` is very dangerous and should be used judiciously).

pelson commented on lib/matplotlib/widgets.py:
```diff
@@ -627,6 +627,9 @@ class RadioButtons(AxesWidget):
      *circles*
         A list of :class:`matplotlib.patches.Circle` instances
+        
+     *val*
```
FWIW - I don't see a good reason not to be very explicit about the name of the variable. ``value`` would be good, ``selected_value`` would be even better...

WeatherGod commented:
Could you initialize the attribute to None at the beginning of the constructor? It is technically possible for the attribute to be undefined if one specifies an "active" value that is outside the range of the list. As for the name of the attribute, there is no need to stick with the other widget's conventions. I personally vote for "selected_label" in order to emphasize that you get back a string, not an index.

As a side note, I will be making a bunch of little PRs tonight, and one of them will be adding a setter method to this class.

WeatherGod commented:
In fact, do we want to go all the way with this and make the attribute a full-fledged property?

DanHickstein commented:
Darn you @tacaswell, first you make me learn how look at the mpl source code, then you make me learn how to use github, then you make me learn this rebase nonsense (which required that I learn how to use vim). Will the learning ever stop?? :)

Anyway, I think that I squashed this to a single commit. I also changed "val" to "value_selected" as suggested. I don't really know what is involved in making this property versus an attribute, but it seems like a good idea to be able to set the status of any of the widgets (for example, someone may want to have a button that resets all of the other buttons to their default values. If I understand correctly, this does not appear to be possible with the current radioButtons.

tacaswell commented:
The value of making it a property rather than an attribute is that you can make it 'read only' by not defining a setter, thus

```
my_buttons.value_selected = 'aardvark'
```
will raise an exception.

If you do define a setter on the property you can validate the assigned value on the way in and update the gui to reflect the changes.

pelson commented:
> The value of making it a property rather than an attribute is that you can make it 'read only' by not defining a setter,

Overkill? There is nothing that is truly read only, and what is the harm in somebody modifying their event's selected_value if they *really* want to do so?

> Could you initialize the attribute to None at the beginning of the constructor?

For me, this is the only critical thing for this PR, but I'm happy if @tacaswell really wants this to become a property.

WeatherGod commented:
Don't worry about making it a property, I can handle that if we want to (my
PR is just about ready, although it isn't done as a property). If we are
happy with the name of the attribute, let's get this merged, and I can then
finish up my PR.

On Thu, Jan 8, 2015 at 5:30 PM, Phil Elson <notifications@github.com> wrote:

> The value of making it a property rather than an attribute is that you can
> make it 'read only' by not defining a setter,
>
> Overkill? There is nothing that is truly read only, and what is the harm
> in somebody modifying their event's selected_value if they *really* want
> to do so?
>
> Could you initialize the attribute to None at the beginning of the
> constructor?
>
> For me, this is the only critical thing for this PR, but I'm happy if
> @tacaswell <https://github.com/tacaswell> really wants this to become a
> property.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/3971#issuecomment-69262015>
> .
>

tacaswell commented:
@pelson I was just babbling about what you can do with properties not saying that it should be done, sorry should have been more clear.

The second usage (the validation) seems more useful anyway as now changing selected and update the GUI as well.

DanHickstein commented:
In the most recent commit, I included the initialization to None at the beginning, so is the current state of the PR fine?

WeatherGod commented:
I am +1 now

tacaswell commented:
@DanHickstein Congratulations on your first mpl contribution! 

Hopefully it is the first of many.

DanHickstein commented:
Awesome! Thanks for walking me through it :)

pelson commented:
> Awesome! Thanks for walking me through it :)

Baptism of fire :fire: :fire_engine: ? :wink: 

Congratulations!
